S3Client fails with exception when executing GetObjectRequest on zero bytes content

### Describe the bug

Performing a getObject operation with an S3Client instance on an existing zero bytes content, after disabling checksum validation, fails with an exception. It happens when using SDK version 2.17.250 and up. It works fine with version 2.17.249.
I guess it has something to do with:
```
## __URL Connection HTTP Client__
  - ### Bugfixes
    - Fix for S3Client with URL Connection http client fails with EOFException when executing HeadObjectRequest for gzip encodeFix to S3Client fails with EOFException.
```

### Expected Behavior

The getObject operation should complete successfully.

### Current Behavior

The call fails with:
```
Exception in thread "main" software.amazon.awssdk.core.exception.SdkClientException: Unable to unmarshall response (abortableInputStream must not be null.). Response Code: 200, Response Text: OK
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:102)
	at software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handleSuccessResponse(CombinedResponseHandler.java:110)
	at software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handleResponse(CombinedResponseHandler.java:75)
	at software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handle(CombinedResponseHandler.java:60)
	at software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handle(CombinedResponseHandler.java:41)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.execute(HandleResponseStage.java:40)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.execute(HandleResponseStage.java:30)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:73)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:78)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:40)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:50)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:81)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:56)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.executeWithTimer(ApiCallTimeoutTrackingStage.java:80)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:60)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:48)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:31)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:37)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:26)
	at software.amazon.awssdk.core.internal.http.AmazonSyncHttpClient$RequestExecutionBuilderImpl.execute(AmazonSyncHttpClient.java:193)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.invoke(BaseSyncClientHandler.java:103)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.doExecute(BaseSyncClientHandler.java:167)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$0(BaseSyncClientHandler.java:68)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:175)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:62)
	at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:52)
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:63)
	at software.amazon.awssdk.services.s3.DefaultS3Client.getObject(DefaultS3Client.java:4483)
	at software.amazon.awssdk.services.s3.S3Client.getObject(S3Client.java:7889)
	at software.amazon.awssdk.services.s3.S3Client.getObject(S3Client.java:8135)
	at com.test.Main.main(Main.java:20)
Caused by: software.amazon.awssdk.core.exception.NonRetryableException: abortableInputStream must not be null.
	at software.amazon.awssdk.core.exception.NonRetryableException$BuilderImpl.build(NonRetryableException.java:92)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler$HttpResponseHandlerAdapter.transformResponse(BaseSyncClientHandler.java:225)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler$HttpResponseHandlerAdapter.handle(BaseSyncClientHandler.java:206)
	at software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handleSuccessResponse(CombinedResponseHandler.java:99)
	... 38 more
Caused by: java.lang.NullPointerException: abortableInputStream must not be null.
	at software.amazon.awssdk.utils.Validate.paramNotNull(Validate.java:156)
	at software.amazon.awssdk.core.ResponseInputStream.<init>(ResponseInputStream.java:43)
	at software.amazon.awssdk.core.sync.ResponseTransformer$1.transform(ResponseTransformer.java:212)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler$HttpResponseHandlerAdapter.transformResponse(BaseSyncClientHandler.java:218)
	... 40 more

Caused by: software.amazon.awssdk.core.exception.NonRetryableException: abortableInputStream must not be null.

Caused by: java.lang.NullPointerException: abortableInputStream must not be null.
```

### Reproduction Steps

```
public class Main {

  static {
    System.setProperty("software.amazon.awssdk.http.service.impl",
      "software.amazon.awssdk.http.urlconnection.UrlConnectionSdkHttpService");
  }

  public static void main(String[] args) {
    String bucket = "bucket";
    String zeroBytesKey = "key";
    try (S3Client client = S3Client.builder()
      .serviceConfiguration(builder -> builder.checksumValidationEnabled(false))
      .build()) {
      client.putObject(b -> b.bucket(bucket).key(zeroBytesKey), RequestBody.empty());
      var length = client.getObject(b -> b.bucket(bucket).key(zeroBytesKey)).response().contentLength();
      System.out.println(length);
    }
  }
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### AWS Java SDK version used

2.17.250

### JDK version used

java 17.0.2 2022-01-18 LTS

### Operating System and version

macOS Monterey Version 12.4

--- a/.changes/next-release/bugfix-AWSSDKforJavav2-a18a347.json
+++ b/.changes/next-release/bugfix-AWSSDKforJavav2-a18a347.json
@@ -1,0 +1,6 @@
+{
+    "type": "bugfix",
+    "category": "AWS SDK for Java v2",
+    "contributor": "",
+    "description": "Fixed an issue in sync clients where empty response payloads could cause a null pointer exception."
+}

--- a/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/handler/BaseSyncClientHandler.java
+++ b/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/handler/BaseSyncClientHandler.java
@@ -203,7 +203,7 @@ public abstract class BaseSyncClientHandler extends BaseClientHandler implements
         @Override
         public ReturnT handle(SdkHttpFullResponse response, ExecutionAttributes executionAttributes) throws Exception {
             OutputT resp = httpResponseHandler.handle(response, executionAttributes);
-            return transformResponse(resp, response.content().orElse(null));
+            return transformResponse(resp, response.content().orElseGet(AbortableInputStream::createEmpty));
         }
 
         @Override

--- a/core/sdk-core/src/main/java/software/amazon/awssdk/core/sync/ResponseTransformer.java
+++ b/core/sdk-core/src/main/java/software/amazon/awssdk/core/sync/ResponseTransformer.java
@@ -172,7 +172,7 @@ public interface ResponseTransformer<ResponseT, ReturnT> {
         return (response, inputStream) -> {
             try {
                 InterruptMonitor.checkInterrupted();
-                return ResponseBytes.fromByteArray(response, IoUtils.toByteArray(inputStream));
+                return ResponseBytes.fromByteArrayUnsafe(response, IoUtils.toByteArray(inputStream));
             } catch (IOException e) {
                 throw RetryableException.builder().message("Failed to read response.").cause(e).build();
             }

--- a/http-client-spi/src/main/java/software/amazon/awssdk/http/AbortableInputStream.java
+++ b/http-client-spi/src/main/java/software/amazon/awssdk/http/AbortableInputStream.java
@@ -17,6 +17,7 @@ package software.amazon.awssdk.http;
 
 import static software.amazon.awssdk.utils.Validate.paramNotNull;
 
+import java.io.ByteArrayInputStream;
 import java.io.FilterInputStream;
 import java.io.InputStream;
 import software.amazon.awssdk.annotations.SdkProtectedApi;
@@ -59,6 +60,10 @@ public final class AbortableInputStream extends FilterInputStream implements Abo
             return new AbortableInputStream(delegate, (Abortable) delegate);
         }
         return new AbortableInputStream(delegate, () -> { });
+    }
+
+    public static AbortableInputStream createEmpty() {
+        return create(new ByteArrayInputStream(new byte[0]));
     }
 
     @Override

--- a/http-clients/url-connection-client/src/it/java/software/amazon/awssdk/http/urlconnection/EmptyFileS3IntegrationTest.java
+++ b/http-clients/url-connection-client/src/it/java/software/amazon/awssdk/http/urlconnection/EmptyFileS3IntegrationTest.java
@@ -1,0 +1,57 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.urlconnection;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static software.amazon.awssdk.testutils.service.S3BucketUtils.temporaryBucketName;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+
+public class EmptyFileS3IntegrationTest extends UrlHttpConnectionS3IntegrationTestBase {
+    private static final String BUCKET = temporaryBucketName(EmptyFileS3IntegrationTest.class);
+
+    @BeforeClass
+    public static void setup() {
+        createBucket(BUCKET);
+    }
+
+    @AfterClass
+    public static void cleanup() {
+        deleteBucketAndAllContents(BUCKET);
+    }
+
+    @Test
+    public void s3EmptyFileGetAsBytesWorksWithoutChecksumValidationEnabled() {
+        try (S3Client s3 = s3ClientBuilder().serviceConfiguration(c -> c.checksumValidationEnabled(false))
+                                            .build()) {
+            s3.putObject(r -> r.bucket(BUCKET).key("x"), RequestBody.empty());
+            assertThat(s3.getObjectAsBytes(r -> r.bucket(BUCKET).key("x")).asUtf8String()).isEmpty();
+        }
+    }
+
+    @Test
+    public void s3EmptyFileContentLengthIsCorrectWithoutChecksumValidationEnabled() {
+        try (S3Client s3 = s3ClientBuilder().serviceConfiguration(c -> c.checksumValidationEnabled(false))
+                                            .build()) {
+            s3.putObject(r -> r.bucket(BUCKET).key("x"), RequestBody.empty());
+            assertThat(s3.getObject(r -> r.bucket(BUCKET).key("x")).response().contentLength()).isEqualTo(0);
+        }
+    }
+}

--- a/http-clients/url-connection-client/src/it/java/software/amazon/awssdk/http/urlconnection/UrlHttpConnectionS3IntegrationTestBase.java
+++ b/http-clients/url-connection-client/src/it/java/software/amazon/awssdk/http/urlconnection/UrlHttpConnectionS3IntegrationTestBase.java
@@ -15,24 +15,12 @@
 
 package software.amazon.awssdk.http.urlconnection;
 
-import static org.assertj.core.api.Assertions.assertThat;
-
 import java.util.Iterator;
 import java.util.List;
 import org.junit.BeforeClass;
-import software.amazon.awssdk.core.ClientType;
-import software.amazon.awssdk.core.interceptor.Context;
-import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
-import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
-import software.amazon.awssdk.http.apache.ApacheHttpClient;
 import software.amazon.awssdk.regions.Region;
-import software.amazon.awssdk.services.s3.S3AsyncClient;
-import software.amazon.awssdk.services.s3.S3AsyncClientBuilder;
 import software.amazon.awssdk.services.s3.S3Client;
 import software.amazon.awssdk.services.s3.S3ClientBuilder;
-import software.amazon.awssdk.services.s3.model.BucketLocationConstraint;
-import software.amazon.awssdk.services.s3.model.CreateBucketConfiguration;
-import software.amazon.awssdk.services.s3.model.CreateBucketRequest;
 import software.amazon.awssdk.services.s3.model.DeleteBucketRequest;
 import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
 import software.amazon.awssdk.services.s3.model.ListObjectVersionsRequest;
@@ -40,7 +28,6 @@ import software.amazon.awssdk.services.s3.model.ListObjectVersionsResponse;
 import software.amazon.awssdk.services.s3.model.ListObjectsRequest;
 import software.amazon.awssdk.services.s3.model.ListObjectsResponse;
 import software.amazon.awssdk.services.s3.model.NoSuchBucketException;
-import software.amazon.awssdk.services.s3.model.S3Exception;
 import software.amazon.awssdk.services.s3.model.S3Object;
 import software.amazon.awssdk.testutils.Waiter;
 import software.amazon.awssdk.testutils.service.AwsTestBase;
@@ -79,6 +66,7 @@ public class UrlHttpConnectionS3IntegrationTestBase extends AwsTestBase {
         Waiter.run(() -> s3.createBucket(r -> r.bucket(bucket)))
               .ignoringException(NoSuchBucketException.class)
               .orFail();
+        s3.waiter().waitUntilBucketExists(r -> r.bucket(bucket));
     }
 
     protected static void deleteBucketAndAllContents(String bucketName) {

